### PR TITLE
Update node-gyp to v10

### LINKED ISSUE #313: Upgrade node-gyp for Python 3.12 support
See also https://github.com/nodejs/node/pull/50209

GitHub Actions will switch to that Python version by default sooner or later.

### PR DESCRIPTION
**Pull request by orgads**
Fixes #313.

### CODE REVIEW

**Comment by orgads** [comment]
Notice this includes also #317

**Comment by orgads** [comment]
@lovell ping. Please also review my other PRs.

File: package.json [hunk 1]
```diff
@@ -68,6 +68,6 @@
   },
   "homepage": "https://github.com/prebuild/prebuild",
   "engines": {
-    "node": "^14.17.0 || ^16.13.0 || >=18.0.0"
+    "node": "^16.13.0 || >=18.0.0"
```
Comment (lovell) [inline package.json:71]: https://github.com/nodejs/node-gyp/pull/2929
```suggestion
    "node": "^16.14.0 || >=18.0.0"
```

File: gyp.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ function runGyp (opts, cb) {
   function runStep () {
     var command = gyp.todo.shift()
     if (!command) {
-      return cb()
+      return cb ? cb() : undefined
```
Comment (lovell) [inline gyp.js:33]: Is `cb` ever undefined? I ask as this function later relies on it... (1/2)

File: gyp.js [hunk 3]
```diff
@@ -40,16 +40,14 @@ function runGyp (opts, cb) {
...
-      }
-
+    gyp.commands[command.name](command.args).then(function () {
       log.verbose('ok')
       process.nextTick(runStep)
+    }, function (err) {
+      log.error(command.name + ' error')
+      log.error('stack', err.stack)
+      log.error('not ok')
+      return cb(err)
```
Comment (lovell) [inline gyp.js:50]: ...here. (2/2)

**Review by lovell — COMMENTED** [review]
Thank you for this PR, I've left a couple of comments inline.

File: gyp.js (hunk 2 above)
Comment (orgads) [reply gyp.js]: In test `gyp is invoked with correct arguments, debug mode`. I added a callback to the test instead.

File: package.json (hunk 1 above)
Comment (orgads) [reply package.json]: Done

**Review by lovell — APPROVED** [review]
Thanks for the updates, LGTM.

File: gyp.js [hunk 4]
```diff
@@ -40,16 +40,14 @@ function runGyp (opts, cb) {
...
 
-    gyp.commands[command.name](command.args, function (err) {
-      if (err) {
-        log.error(command.name + ' error')
-        log.error('stack', err.stack)
-        log.error('not ok')
-        return cb(err)
-      }
-
+    gyp.commands[command.name](command.args).then(function () {
```
Comment (mathiask88) [inline gyp.js:43]: @orgads @lovell Does this change also work with other backends like node-ninja? AFAIK they are still running callback style functions.

Comment (lovell) [reply gyp.js:43]: https://github.com/prebuild/prebuild/issues/323

**Pull request merged**